cirq-core/cirq/testing/sample_gates.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2023 The Cirq Developers
```

Comment:
My proposal for this file was to have a library of well defined gates which cover different cases of ancilla allocation; eg:

1) Allocate ancilla but expect them to be clean and leave them in a clean state (eg: `GateThatAllocatesAQubit`) 
2) Allocate ancilla and expect them to be in any dirty state and leave them in the same state (eg: `GateThatAllocatesTwoQubits`)


For such gates; we should have documentation and tests for their decomposition as part of `cirq-core/cirq/testing/sample_gates_test.py`. The goal is to make these gates reusable for other tests (like reduced unitary / simulations). 

The more simpler cases like `NotDecomposableGate` or more specific cases like `DecomposableGate` which doesn't have any guarantees on the can (and probably should) live in specific test files. 

We can also do this as part of a follow-up PR if you'd prefer; but this was my main motivation for suggesting that we create a separate file in `cirq/testing`

We need to add documentation and tests for gates in this file. The idea is to keep things like `GateThatAllocatesTwoQubits` in this file; which essentially has a decomposition where we can borrow two qubits in any state (with a docstring that clarifies it and a test which ensures that the decomposition is correct).

Reply:
Lets keep the interesting gates here as is and I'll improve their docstring and add _test in a followup PR

Reply:
done